src/formatter.js

```diff
@@ -55,81 +55,6 @@ function lookAhead(notes, restLine, i, compare) {
   return nextRestLine;
 }
 
-// Take an array of `voices` and place aligned tickables in the same context. Returns
-// a mapping from `tick` to `ContextType`, a list of `tick`s, and the resolution
-// multiplier.
-//
-// Params:
-// * `voices`: Array of `Voice` instances.
-// * `ContextType`: A context class (e.g., `ModifierContext`, `TickContext`)
-// * `addToContext`: Function to add tickable to context.
-function createContexts(voices, ContextType, addToContext) {
-  if (!voices || !voices.length) {
-    throw new Vex.RERR('BadArgument', 'No voices to format');
-  }
-
-  // Find out highest common multiple of resolution multipliers.
-  // The purpose of this is to find out a common denominator
-  // for all fractional tick values in all tickables of all voices,
-  // so that the values can be expanded and the numerator used
-  // as an integer tick value.
-  const totalTicks = voices[0].getTotalTicks();
-  const resolutionMultiplier = voices.reduce((resolutionMultiplier, voice) => {
-    if (!voice.getTotalTicks().equals(totalTicks)) {
-      throw new Vex.RERR('TickMismatch', 'Voices should have same total note duration in ticks.');
-    }
-
-    if (voice.getMode() === Voice.Mode.STRICT && !voice.isComplete()) {
-      throw new Vex.RERR('IncompleteVoice', 'Voice does not have enough notes.');
-    }
-
-    return Math.max(resolutionMultiplier, Fraction.LCM(resolutionMultiplier, voice.getResolutionMultiplier()));
-  }, 1);
-
-  // Initialize tick maps.
-  const tickToContextMap = {};
-  const tickList = [];
-  const contexts = [];
-
-  // For each voice, extract notes and create a context for every
-  // new tick that hasn't been seen before.
-  voices.forEach((voice, voiceIndex) => {
-    // Use resolution multiplier as denominator to expand ticks
-    // to suitable integer values, so that no additional expansion
-    // of fractional tick values is needed.
-    const ticksUsed = new Fraction(0, resolutionMultiplier);
-
-    voice.getTickables().forEach((tickable) => {
-      const integerTicks = ticksUsed.numerator;
-
-      // If we have no tick context for this tick, create one.
-      if (!tickToContextMap[integerTicks]) {
-        const newContext = new ContextType({ tickID: integerTicks });
-        contexts.push(newContext);
-        tickToContextMap[integerTicks] = newContext;
-      }
-
-      // Add this tickable to the TickContext.
-      addToContext(tickable, tickToContextMap[integerTicks], voiceIndex);
-
-      // Maintain a sorted list of tick contexts.
-      tickList.push(integerTicks);
-      ticksUsed.add(tickable.getTicks());
-    });
-  });
-
-  return {
-    map: tickToContextMap,
-    array: contexts,
-    list: Vex.SortAndUnique(
-      tickList,
-      (a, b) => a - b,
-      (a, b) => a === b
-    ),
-    resolutionMultiplier,
-  };
```

Comment:
For this, I would rather keep the code as it is today and add new simple wrapper functions: `createTickContext` and `createModifierContexts` that call this with the right types. I really do not want to unwind this code. :-)

Reply:
I mean something like this

```javascript
createTickContexts(voices: Voice[]) {
  return  createContexts(voices, TickContext, ...);
}

createModifierContexts(voices Voice[]) {
  return createContexts(voices, ModifierContext, ...);
}
```

Reply:
This can keep the call sites clean, but preserve the shared implementation.

Reply:
Did you look at this diff? 

```diff
-  // Create `ModifierContext`s for each tick in `voices`.
-  createModifierContexts(voices) {
+  // Create `TickContext`s for each tick in `voices`. Also calculate the
+  // total number of ticks in voices.
+  createTickContexts(voices) {
     const resolutionMultiplier = Formatter.getResolutionMultiplier(voices);
     // Initialize tick maps.
     const tickToContextMap = {};
     const tickList = [];
     const contexts = [];
 
     // For each voice, extract notes and create a context for every
     // new tick that hasn't been seen before.
-    voices.forEach((voice) => {
+    voices.forEach((voice, voiceIndex) => {
       // Use resolution multiplier as denominator so that no additional expansion
       // of fractional tick values is needed.
       const ticksUsed = new Fraction(0, resolutionMultiplier);

       voice.getTickables().forEach((tickable) => {
         const integerTicks = ticksUsed.numerator;
 
-        // If we have no modifier context for this tick, create one.
+        // If we have no tick context for this tick, create one.
         if (!tickToContextMap[integerTicks]) {
-          const newContext = new ModifierContext();
+          const newContext = new TickContext({ tickID: integerTicks });
           contexts.push(newContext);
           tickToContextMap[integerTicks] = newContext;
           // Maintain a list of unique integerTicks.
           tickList.push(integerTicks);
         }
 
-        // Add this ModifierContext to the tickable.
-        tickable.addToModifierContext(tickToContextMap[integerTicks]);
+        // Add this tickable to the TickContext.
+        tickToContextMap[integerTicks].addTickable(tickable, voiceIndex);
         ticksUsed.add(tickable.getTicks());
       });
     });
 
-    this.modiferContexts = {
+    contexts.forEach((context) => {
+      context.tContexts = contexts;
+    });
+
+    this.totalTicks = voices[0].getTicksUsed().clone();
+    this.tickContexts = {
       map: tickToContextMap,
       array: contexts,
       list: tickList.sort((a, b) => a - b),
       resolutionMultiplier,
     };
-    return this.modiferContexts;
+    return this.tickContexts;
   }
```

Reply:
> I mean something like this:
> 
> ```js
> createTickContexts(voices: Voice[]): TickContext {
>   return createContexts(voices, TickContext, ...);
> }
> 
> createModifierContexts(voices Voice[]): ModifierContext {
>   return createContexts(voices, ModifierContext, ...);
> }
> ```

The problem is that createContext would be like that:
```js
   createContexts(voices, ModifierContext, ...): TickContext | ModierContext {
```
And we cannot just return them in the functions above.

Reply:
In fact more complex
```
function createContexts(
   voices: Voice[],
   ContextType: typeof ModifierContext|typeof TickContext,
   addToContext: Function (tickable: Tickable, context: ModifierContext):void | Function (tickable: Tickable, context: TickableContext, voiceId: number):void
) : ModifierContextsStruct | TickContextsStruct

And this function expandes to `ModifierContext({ tickID: integerTicks })` which is not correct.

@I have closed all the other conversations in this PR leaving the important information in this one. 
I will move to another topic and leave formatter free in progress list do that others can take over.

Reply:
I think you could do something like this?

```javascript
type ContextType = ModifierContext | TickContext

interface AlignmentContexts {
    map: Record<number, ContextType>,
    array: ContextType[],
    list: ContextType[]
    resolutionMultiplier: number
}

type addToContextFn = (tickable: Tickable, context: ContextType, voiceIndex?: number) => void
type makeContextFn = (tick?: {tickID: number}) => ContextType

// This is the original createContexts: we update params and types and the call to makeContext inside.
function createContexts(voices: string[], makeContext: makeContextFn, addToContext: addToContextFn): AlignmentContexts {
    // add types
    // replace 'new ModifierContext(...)/TickContext(...)' with: 'makeContext({tickID: ...})';
    return {
        map: ...,
        array: ...,
        list: ...,
        resolutionMultiplier,
    };
}

// These are the wrappers to use in the call sites
function createModifierContexts(voices: string[]): AlignmentContexts {
    const fn: addToContextFn = (tickable: Tickable, context: ContextType, voiceIndex?: number) => tickable.addToModifierContext(context);
    return createContexts(voices, () => new ModifierContext(), fn);
}

function createTickContexts(voices: string[]): AlignmentContexts {
    const fn: addToContextFn = (tickable: Tickable, context: ContextType, voiceIndex?: number) => (context as TickContext).addTickable(voiceIndex as number);
    return createContexts(voices, (tickID) => new TickContext(tickID), fn);
}
```

Reply:
I believe that this will complain about `tickable` not being used:
```js
const fn: addToContextFn = (tickable: Tickable, context: ContextType, voiceIndex?: number) => (context as TickContext).addTickable(voiceIndex as number);
 ```